File: ibkr_show_backend/app/schemas/financial_calendar.py
from pydantic import BaseModel


class EarningsEvent(BaseModel):
    symbol: str = ""
    name: str = ""
    mcap: float = 0
    exchange: str = ""
    date: str = ""
    is_estimate: bool = True
    eps_avg: float | None = None
    eps_low: float | None = None
    eps_high: float | None = None
    rev_avg: float | None = None
    rev_low: float | None = None
    rev_high: float | None = None
    call_date: str = ""


class EarningsCalendarResponse(BaseModel):
    items: list[EarningsEvent]


class MacroEvent(BaseModel):
    date: str = ""
    title: str = ""
    type: str = "other"
    importance: str = "low"
    description: str = ""
    time: str | None = None
    actual: str | None = None
    expected: str | None = None
    previous: str | None = None


class MacroEventResponse(BaseModel):
    items: list[MacroEvent]

File: ibkr_show_frontend/src/api/financialCalendar.ts
import { request } from './http'
import { withCache } from '@/utils/cache'

export interface EarningsEvent {
  symbol: string
  name: string
  mcap: number
  exchange: string
  date: string
  is_estimate: boolean
  eps_avg: number | null
  eps_low: number | null
  eps_high: number | null
  rev_avg: number | null
  rev_low: number | null
  rev_high: number | null
  call_date: string
}

export interface EarningsCalendarResponse {
  items: EarningsEvent[]
}

function toQueryString(params: Record<string, string | number | undefined | null>): string {
  const searchParams = new URLSearchParams()
  Object.entries(params).forEach(([key, value]) => {
    if (value !== undefined && value !== null && value !== '') {
      searchParams.set(key, String(value))
    }
  })
  const queryString = searchParams.toString()
  return queryString ? `?${queryString}` : ''
}

export function fetchEarningsCalendar(startDate: string, endDate: string): Promise<EarningsCalendarResponse> {
  return withCache(
    `earnings_${startDate}_${endDate}`,
    () => request<EarningsCalendarResponse>(`/api/financial-calendar/earnings${toQueryString({ start_date: startDate, end_date: endDate })}`)
  )
}
